pennylane/resource/estimation.py

```diff
@@ -0,0 +1,203 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for resource estimation.
+"""
+
+from collections import defaultdict
+import pennylane as qml
+from pennylane import numpy as np
+from pennylane.operation import AnyWires, Operation
+
+
+class Resource:
+    r"""Create a resource object for storing quantum resource information."""
+
+    def __init__(self):
+        self.num_wires = 0
+        self.num_gates = 0
+        self.gate_types = defaultdict(int)
+        self.depth = 0
+        self.shots = 0
+
+    def __str__(self):
+        keys = ["wires", "gates", "depth", "shots", "gate_types"]
+        vals = [self.num_wires, self.num_gates, self.depth, self.shots, self.gate_types]
+        items = "\n".join([str(i) for i in zip(keys, vals)])
+        items = items.replace("('", "")
+        items = items.replace("',", ":")
+        items = items.replace(")", "")
+        items = items.replace("defaultdict(<class 'int'>, ", "\n")
+        return items
+
+    def __repr__(self):
+        return f"<Resource: wires={self.num_wires}, gates={self.num_gates}, depth={self.depth}, shots={self.shots}, gate_types={self.gate_types}>"
+
+    def _ipython_display_(self):
+        """Displays __str__ in ipython instead of __repr__"""
+        print(str(self))
+
+
+class CustomOperation(qml.DoubleExcitation):
+    num_wires = 4
+
+    def resources(self):
+        resource = Resource()
+        resource.num_gates = 11
+        resource.gate_types = {"T": 5, "CNOT": 6}
+        resource.depth = 11
+        return resource
+
+
+def estimate_resources_tape(tape):
```

Comment:
@rmoyard - this is another data point for a use-case of transforms.

Here we want to input a tape and return a non-tape (`Resources`) object.